### ISSUE TITLE
haproxy config

### ISSUE THREAD

**Issue by knfoo** [kubernetes-mail-server/postfix#2]
Hi,

I have looked at mail setup and looks very nice.
You are referring to haproxy but I cannot find the configuration.
Would you be able to share this ?

**Comment by christhomas** [comment]
HAProxy Protocol is just a way for a load balancer to pass through the original IP address to an internal service. In this case, Kubernetes Ingress Controller is the ingress where connections from the outside world enter the cluster. 

When this ingress controller routes the request through to postfix, postfix checks the IP address against it's filtering rules and because it's an internal IP address, it always succeeds, leaving your postfix as an "Open Relay"

This is because postfix uses the connecting IP Address to go through all the restriction rules and checking the fqdn and all sorts of things, but in particular "permit_mynetworks" will ALWAYS match your ingress controllers ip address, since they're all on the same network (probably). 

Therefore EVERY connection postfix gets, is trusted because all it sees is the ingress controller.

The solution is to use this postscreen setup and use the haproxy protocol. Kubernetes Ingress Controller supports using the proxy protocol and attaches the external IP Address to the request and passes it along to postfix.

Thing is, natively postfix doesn't support haproxy protocol and that's why I setup postscreen in this way to decode and "rewrite" the request with the correct EXTERNAL IP address.

Then, invalid mail servers that try to use your server to send email (an open relay, spam relay, whatever), all get denied access and then this closes the security loophole and in theory, everybody is happy.

You can also checkout this other repository which has the information on how my cluster is setup. In particular, read the ingress manifest, you'll find where I'm setting up the ingress for the mail server ports: https://github.com/christhomas/kubernetes-cluster

**Comment by christhomas** [comment]
Btw, this setup does work very well and I'm using it for my daily personal domain. I've not had to update or restart it in months. It seems very stable and I'm open to suggestions if you're experienced with postfix and want to make any.

**Comment by knfoo** [comment]
Thank you for the explanation on how the connection between ingress and the mail setup works.
And thanks for pointing me to your kubernetes ingress config ;)https://github.com/christhomas/kubernetes-cluster/blob/master/02-ingress-controller.yaml
contains the missing link between the proxy protocol and the ingress that I was after :)
To get the setup to work as expected. 
Last question do you have a public ip pointing directly to your ingress node or do you have nat in front of the cluster ?
I do not have any modifications atm but will let you know if I find any ;) great work - love that you have gitlab ci/cd and the `.gitlab-ci.yml` files in the repos :+1:

**Comment by christhomas** [comment]
The domain is registered and I've configured the DNS Zone to have multiple A records pointing to each IP Address of each VPS node in the cluster.

I have three VPS nodes, all setup with the kubernetes-cluster repository I linked you to before, all clustered together.

The ingress is configured as a DaemonSet, meaning (if you didn't already know) that it'll run the Ingress Controller on each node in the cluster. So three nodes, three ingress controller pods, all using HostPort directly onto the VPS nodes ports. Meaning all my port 80's on every node are "forwarded" into the cluster.

This allows me to take a generic VPS, install ONLY docker + kubernetes, then run all my services and software on the cluster and define which nodes certain software run using NodeSelectors to force particular pods onto particular nodes.

This is important because I use the HostPath provisioner to mount VPS local hard disk into the cluster and use it for permanent storage. This is what I do to store the database and mail server contents. It's all local hard disk that's "mounted" into the cluster. See in the kubernetes cluster for how I setup the host path provisioner.

Then in the mail server, I create persistent storage for each "drive" that I want to store data to and then configure the pods to use them. I didn't yet create any cluster storage devices which would mean any pod on any node can access all data from anywhere. This is something I want to do in the future. I'm thinking to use Rook to do this.

But until then, I have to mount local disks into the cluster. This means that dovecot, which is responsible for storing the mail data, needs to execute on a particular node and I have to use Node Selectors to lock that pod onto the right node, so it mounts the correct file system and the contents are there. Because I have three nodes and I only use one VPS node for storing mail server data.

So this part is a bit tricky. But it works.

TL;DR: yes, I configure the dns records to point to the nodes :)

The gitlab-ci.yml files are because I was originally using this pipeline to deploy everything. So it's not really maintained anymore, as in I didn't really check they work 100%, since I changed to using github and travis for deployment. It might be that they are out of date. If you notice anything, please let me know.

But then since I wanted to make the project public, I swapped to using github,  created travis pipelines and then I removed all the gitlab pipelines.

**Comment by christhomas** [comment]
If you like gitlab, you might want to checkout my kubernetes-gitlab repository. This might be interesting to you. Although tbh, gitlab is fucking heavy. I'm thinking to change to gitea. I have a repository for this software too. But it's really not usable right now.  

The gitlab one I'm using for real repositories. You could say it's "Production", at least at the scale that I'm using it :)

https://github.com/christhomas/kubernetes-gitlab

**Comment by christhomas** [comment]
ok, I guess you can always reopen this issue if you have any further questions

**Comment by knfoo** [comment]
I have a nice working gitlab installation with my k8s setup but thanks.
It seems that I have got my haproxy ingress to work now which is great.
Once I have the setup finalized I will share my config and changes :)

I have a question. 
How do you manage users, domains, aliases, opendkim in this setup ?
Do you use something like postadmin or how does it work ?

**Comment by knfoo** [comment]
As far as I can tell I cannot re-open this as I am not a collaborator in this project and you closed it.

**Comment by christhomas** [comment]
If you look at my setup you'll see that my opendkim is configured by database.

That's where my setup gets a bit tricky because the API to add and remove 
let's and users isn't built yet. But the basic setup is there when I get 
time to build it

Regards
Chris

**Comment by christhomas** [comment]
It would be nice to see your gitlab setup if you have it on GitHub 
someplace. Perhaps I find something that if like to copy :)

**Comment by knfoo** [comment]
I will share my gitlab stuff once I have the setup running in a working state.
It is kind of in my spare time that I am doing it so it takes a while.

**Comment by knfoo** [comment]
I will share my gitlab stuff once I have the setup running in a working state.
It is kind of in my spare time that I am doing it so it takes a while.
